dewy/common/collection_embeddings.py

```diff
@@ -0,0 +1,259 @@
+        # of embeddings.
+        dimensions = get_dimensions(self.text_embedding_model)
+        field = f"embedding::vector({dimensions})"
+
+        self._retrieve_embeddings = f"""
+        SELECT
+          chunk_id,
+          {self.text_distance_metric.distance(field, "$1")} AS score
+        FROM embedding
+        WHERE collection_id = {self.collection_id}
+        ORDER BY {self.text_distance_metric.order_by(field, "$1")}
+        LIMIT $2
+        """
+
+        self._retrieve_chunks = f"""
+        WITH relevant_embeddings AS (
+          SELECT
+            chunk_id,
+            {self.text_distance_metric.distance(field, "$1")} AS score
+          FROM embedding
+          WHERE collection_id = {self.collection_id}
+          ORDER BY {self.text_distance_metric.order_by(field, "$1")}
+        )
+        SELECT
+          relevant_embeddings.chunk_id AS chunk_id,
+          chunk.text AS text,
+          relevant_embeddings.score AS score
+        FROM relevant_embeddings
+        JOIN chunk
+        ON chunk.id = relevant_embeddings.chunk_id
+        LIMIT $2
+        """
+
+    @staticmethod
+    async def for_collection_id(pg_pool: asyncpg.Pool, collection_id: int) -> Self:
+        """Retrieve the collection embeddings of the given collection."""
+        async with pg_pool.acquire() as conn:
+            result = await conn.fetchrow(
+                """
+                SELECT
+                    name,
+                    id as collection_id,
+                    text_embedding_model,
+                    text_distance_metric
+                FROM collection
+                WHERE collection.id = $1;
+                """,
+                collection_id,
+            )
+
+            return CollectionEmbeddings(pg_pool, result)
+
+    @staticmethod
+    async def for_document_id(pg_pool: asyncpg.Pool, document_id: int) -> (str, Self):
+        """Retrieve the collection embeddings and the URL of the given document."""
+
+        # TODO: Ideally the collection embeddings would be cached, and this
+        # wouldn't need to exist.
+        async with pg_pool.acquire() as conn:
+            result = await conn.fetchrow(
+                """
+                SELECT
+                    document.id,
+                    document.url,
+                    collection.name,
+                    collection.id as collection_id,
+                    collection.text_embedding_model,
+                    collection.text_distance_metric
+                FROM document
+                JOIN collection ON document.collection_id = collection.id
+                WHERE document.id = $1;
+                """,
+                document_id,
+            )
+
+            # TODO: Cache the configured ingestions, and only recreate when needed?
+            configured_ingestion = CollectionEmbeddings(pg_pool, result)
+            return (result["url"], configured_ingestion)
+
+    async def retrieve_text_embeddings(
```

Comment:
Where is this used?

Reply:
My intention here is to have a "retrieve the embeddings without grouping them by chunk" for diagnostic purposes. It's not currently used.